core/modules/drr.cc

```diff
@@ -0,0 +1,392 @@
+    // checks to see if there has been no update after a full round
+    if (count == 0) {
+      if (batch_size == batch->cnt()) {
+        break;
+      } else {
+        count = llring_count(flow_ring_);
+        batch_size = batch->cnt();
+      }
+    }
+    count--;
+
+    // gets the next flow in the round
+    if (!current_flow_) {
+      err =  llring_sc_dequeue(flow_ring_, reinterpret_cast<void**>(&f));
+      if (err < 0)
+        return err;
+
+      if (llring_empty(f->queue) && !f->next_packet) {
+        // if the flow expired, remove it
+        if (now - f->timer > kTTL) {
+          RemoveFlow(f);
+        } else {
+          err = llring_sp_enqueue(flow_ring_, f);
+          if (err < 0)
+            return err;
+        }
+        continue;
+      }
+
+      f->deficit += quantum_;
+    } else {
+      f = current_flow_;
+      current_flow_ = nullptr;
+    }
+
+    int bytes = GetNextPackets(batch, f);
+    if (bytes < 0) {
+      return bytes;
+    }
+    total_bytes += bytes;
+
+    if (llring_empty(f->queue) && !f->next_packet) {
+      f->deficit = 0;
+    }
+
+    // if the flow doesn't have any more packets to give, reenqueue it
+    if (!f->next_packet || f->next_packet->total_len() > f->deficit) {
+      err = llring_sp_enqueue(flow_ring_, f);
+      if (err < 0)
+        return err;
+    } else  {
+      // knowing that the while statement will exit, keep the flow that still
+      // has packets at the front
+      current_flow_ = f;
+    }
+  }
+  return total_bytes;
+}
+
+int DRR::GetNextPackets(bess::PacketBatch* batch, Flow* f) {
+  int total_bytes = 0;
+  bess::Packet* pkt;
+
+  while (!batch->full() && !llring_empty(f->queue)) {
+    // makes sure there isn't already a packet at the front
+    if (!f->next_packet) {
+      int err = llring_sc_dequeue(f->queue, reinterpret_cast<void**>(&pkt));
+      if (err < 0)
+        return err;
+    } else {
+      pkt = f->next_packet;
+      f->next_packet = nullptr;
+    }
+
+    if (pkt->total_len() > f->deficit) {
+      f->next_packet = pkt;
+      break;
+    }
+
+    f->deficit -= pkt->total_len();
+    total_bytes += pkt->total_len();
+    batch->add(pkt);
+  }
+
+  return total_bytes;
+}
+
+DRR::Flow_Id DRR::GetId(bess::Packet* pkt) {
+  struct ether_hdr *eth = pkt->head_data<struct ether_hdr *>();
+  struct ipv4_hdr *ip = reinterpret_cast<struct ipv4_hdr *>(eth + 1);
+  int ip_bytes = (ip->version_ihl & 0xf) << 2;
+  struct udp_hdr *udp = reinterpret_cast<struct udp_hdr *>(
+      reinterpret_cast<uint8_t *>(ip) + ip_bytes);  // Assumes a l-4 header
+  // TODO(joshua): handle packet fragmentation
+  Flow_Id id = {ip->src_addr, ip->dst_addr, udp->src_port,
+      udp->dst_port, ip->next_proto_id};
+  return id;
+}
+
+int DRR::AddNewFlow(bess::Packet *pkt, Flow_Id id) {
+  int err;
+
+  // creates flow
+  Flow* f = static_cast<Flow *>(mem_alloc_ex(sizeof(Flow), alignof(Flow), 0));
+  Flow init = {0, 0, id, 0, nullptr};
+  memcpy(f, &init, sizeof(init));
```

Comment:
Is there a reason you use mem_alloc here rather than new?  You could define a constructor on Flow then this would just be Flow *f = new Flow().